graphql_compiler/macros/macro_edge/validation.py

```diff
@@ -0,0 +1,249 @@
+                existing_typename_field = copy(selection)
+                existing_typename_field.directives = copy(existing_typename_field.directives)
+                existing_typename_field.directives.append(output_directive)
+                first_level_selections[idx] = existing_typename_field
+    if existing_typename_field is None:
+        first_level_selections.insert(0, Field(Name('__typename'), directives=[output_directive]))
+
+    # Propagate the changes back to the result_ast
+    root_level_selection.selection_set = SelectionSet(first_level_selections)
+    query_ast.selection_set = SelectionSet([root_level_selection])
+    return Document([query_ast])
+
+
+# ############
+# Public API #
+# ############
+
+def get_and_validate_macro_edge_info(schema, ast, macro_edge_args,
+                                     type_equivalence_hints=None):
+    """Return a tuple with the three parts of information that uniquely describe a macro edge.
+
+    Args:
+        schema: GraphQL schema object, created using the GraphQL library
+        ast: GraphQL library AST OperationDefinition object, describing the GraphQL that is defining
+             the macro edge.
+        macro_edge_args: dict mapping strings to any type, containing any arguments the macro edge
+                         requires in order to function.
+        type_equivalence_hints: optional dict of GraphQL interface or type -> GraphQL union.
+                                Used as a workaround for GraphQL's lack of support for
+                                inheritance across "types" (i.e. non-interfaces), as well as a
+                                workaround for Gremlin's total lack of inheritance-awareness.
+                                The key-value pairs in the dict specify that the "key" type
+                                is equivalent to the "value" type, i.e. that the GraphQL type or
+                                interface in the key is the most-derived common supertype
+                                of every GraphQL type in the "value" GraphQL union.
+                                Recursive expansion of type equivalence hints is not performed,
+                                and only type-level correctness of this argument is enforced.
+                                See README.md for more details on everything this parameter does.
+                                *****
+                                Be very careful with this option, as bad input here will
+                                lead to incorrect output queries being generated.
+                                *****
+
+    Returns:
+        tuple (class name for macro, name of macro edge, MacroEdgeDescriptor),
+        where the first two values are strings and the last one is a MacroEdgeDescriptor object
+    """
+    macro_directives = get_directives_for_ast(ast)
+
+    _validate_macro_ast_with_macro_directives(schema, ast, macro_directives)
```

Comment:
I think this function could be much more readable if the validation helper functions would have better names. I recommend a number of ways to change these functions so that it is more clear what is going on.